rak811.go

```diff
@@ -147,180 +167,155 @@ func (l *Lora) Close() {
+func (l *Lora) JoinOTAA() (string, error) {
+	resp, err := l.txr("join=otaa", 1)
 	if err != nil {
 		return "", err
 	}
 	if resp != OK {
 		return "", errors.Errorf("invalid join request response resp:%v", resp)
 	}
-	start := time.Now()
-	reader := bufio.NewReader(l.port)
-	for {
-		resp, err = reader.ReadString('\n')
-		resp = strings.TrimSuffix(strings.TrimSpace(resp), "\r")
-		if err == nil && resp != "" {
-			switch resp {
-			case JoinSuccess:
-				return JoinSuccess, nil
-			case JoinFail:
-				return JoinFail, nil
-			case JoinTimeout:
-				return JoinTimeout, nil
-			default:
-				return "", errors.Errorf("invalid join response resp:%v", resp)
-			}
-		}
-		if time.Since(start) > timeout {
-			return "", errors.New("no response after the set timeout")
-		}
-	}
+	// The module doesn't accept any other command before it returns a response
+	// so need to wait for it.
+	return l.tr(1)
 }
 
-// JoinABP join the configured network in ABP mode
+// JoinABP join the configured network in ABP mode.
 func (l *Lora) JoinABP() (string, error) {
-	return l.tx("join=abp")
+	return l.txr("join=abp", 1)
 }
 
-// Signal check the radio rssi, snr, update by latest received radio packet
+// Signal check the radio rssi, snr, update by latest received radio packet.
 func (l *Lora) Signal() (string, error) {
-	return l.tx("signal")
+	return l.txr("signal", 1)
 }
 
-// GetDataRate get next send data rate
+// GetDataRate get next send data rate.
 func (l *Lora) GetDataRate() (string, error) {
-	return l.tx("dr")
+	return l.txr("dr", 1)
 }
 
-// SetDataRate set next send data rate
+// SetDataRate set next send data rate.
 func (l *Lora) SetDataRate(datarate string) (string, error) {
-	return l.tx(fmt.Sprintf("dr=%s", datarate))
+	return l.txr(fmt.Sprintf("dr=%s", datarate), 1)
 }
 
-// GetLinkCnt get LoRaWAN uplink and downlink counter
+// GetLinkCnt get LoRaWAN uplink and downlink counter.
 func (l *Lora) GetLinkCnt() (string, error) {
-	return l.tx("link_cnt")
+	return l.txr("link_cnt", 1)
 }
 
-// SetLinkCnt set LoRaWAN uplink and downlink counter
+// SetLinkCnt set LoRaWAN uplink and downlink counter.
 func (l *Lora) SetLinkCnt(uplinkCnt, downlinkCnt float32) (string, error) {
-	return l.tx(fmt.Sprintf("link_cnt=%f,%f", uplinkCnt, downlinkCnt))
+	return l.txr(fmt.Sprintf("link_cnt=%f,%f", uplinkCnt, downlinkCnt), 1)
 }
 
 // GetABPInfo
 func (l *Lora) GetABPInfo() (string, error) {
-	return l.tx("abp_info")
+	return l.txr("abp_info", 1)
 }
 
-// Send send data to LoRaWAN network
+// Send send data to LoRaWAN network.
+// The parent function should check for:
+// STATUS_TX_UNCOMFIRMED when sending unconfirmed packets,
+// STATUS_TX_COMFIRMED when sending confirmed packets.
 func (l *Lora) Send(data string) (string, error) {
-	resp, err := l.tx(fmt.Sprintf("send=%s", data))
+	resp, err := l.txr(fmt.Sprintf("send=%s", data), 1)
 	if err != nil {
 		return "", err
 	}
 	if resp != OK {
```

Comment:
In case is not an `OK` response, we need to return the error response because it contains information about the error and flag the error of course.

On the formatting side of things leave the space after the colon.

Reply:
yep, good point, modifier to just return the resp as an error